### PR TITLE
[Bug][RL] Zero-valid-token batches still step optimizer and advance policy version

### PR DESCRIPTION
**Pull request by taking-lying-flat**
## Problem

A generated RL training sample can contain tokens while still contributing no learning signal after the causal shift. DAPO only trains a shifted token when loss_mask is true and the corresponding generator log probability is finite. If every shifted token fails either condition, the effective loss mask is empty.

Previously these samples still made a rollout group look trainable. The group could count toward num_prompts_per_train_step, produce a batch, run optimizer.step, and advance the policy version even though it contributed no valid loss token. Optimizer state, weight decay, and version synchronization mean that this is not necessarily a harmless no-op.

There is a related active-slot leak after batching. TrainingSampleBuilder can return a non-empty group, but Batcher may then drop every sample because it exceeds seq_len. The old add_training_samples return value used None both when the batcher was merely accumulating valid groups and when the current group became empty. The controller therefore could not know that it needed to release the emptied group's active rollout slot; repeated cases could exhaust the active window and stall the pipeline.

## Root cause

Trainability was inferred from whether a group contained TrainingSample objects rather than whether those samples had any shifted token usable by the loss. In addition, the Batcher API did not expose whether its own oversized-sample filtering had made the newly added group untrainable.

## Fix

- Filter each zero-valid-token sample in TrainingSampleBuilder, where loss_mask and generator log probabilities are produced and data validity belongs.
- Keep oversized-sample filtering in Batcher because it depends on batch shape and seq_len.
- Change the existing Batcher.add_training_samples method to return both the optional TrainingBatch and a group_is_trainable boolean.
- Let the controller release the active group immediately when Batcher reports false.
- Keep metric-only groups available for metric aggregation without counting them toward a train step.

The latest review request is implemented directly: there is no add_training_samples_with_status helper or compatibility wrapper. The original add_training_samples API now returns the tuple, and its controller and unit-test call sites are updated.

The branch is merged with the latest main, including the fold-batch-dim change.

### CODE REVIEW

**Comment by tianyu-l** [comment]
@claude what problem is this PR trying to address? Could you give an example?

**Comment by tianyu-l** [comment]
@claude I now understand the issue. However, compared with changes in this PR, I think the better place to put such filter is in https://github.com/pytorch/torchtitan/blob/main/torchtitan/experiments/rl/components/training_sample_builder.py#L43, which according to the docstring, is meant to "Apply group-level filters and convert surviving rollouts to training_samples." What do you think?

**Comment by tianyu-l** [comment]
@claude 

> The builder sees a valid token (from X) and keeps the group. The batcher then drops X as oversized, leaving Y — now zero-valid — which sails through as "trainable" and reintroduces exactly the zero-gradient optimizer step this PR fixes.

I see this subtlety. I think we can modify `TrainingSampleBuilder` so that `build_from_group` or even `rollout_to_training_samples` directly filters out sample Y. The numerics shouldn't be affected, and we can continue to filter X by the batcher so that batcher doesn't need to worry about zero-valid cases.

**Comment by tianyu-l** [comment]
@taking-lying-flat please see if my latest comment makes sense

File: torchtitan/experiments/rl/controller.py [hunk 1]
```diff
@@ -1007,6 +1007,10 @@ async def _batcher_loop(
                     rollout_group=rollout_group
                 )
 
+            training_sample_group = await asyncio.to_thread(
+                batcher.prepare_training_sample_group,
+                training_sample_group=training_sample_group,
+            )
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/controller.py]: can we delete this and the introduction of `prepare_training_sample_group` now?

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/experiments/rl/controller.py (hunk 1 above)
Comment (taking-lying-flat) [reply torchtitan/experiments/rl/controller.py]: Thanks — I removed the controller pre-pass and the prepare_training_sample_group entry point as requested. The next RL integration run then exposed a separate interaction with the existing oversized-sample filter: the job timed out after Batcher logged that it dropped 2/2 samples exceeding seq_len=1024. A group can be non-empty after TrainingSampleBuilder filtering, pass the controller check, and then become empty inside Batcher when all remaining samples are oversized. Because add_training_samples previously returned only None for both an accumulating batch and a group emptied by filtering, the controller could not release that group active slot. Repeated occurrences exhausted the active window, leaving the producer waiting for a slot and the trainer waiting for a batch. The latest change keeps the requested single-pass structure: Batcher now atomically adds and filters the group while also reporting whether the group remains trainable, and the controller releases the slot from that post-filter status. The original add_training_samples API remains as a compatibility wrapper, so there is no preparation pre-pass or duplicated filtering.

File: torchtitan/experiments/rl/components/batcher.py [hunk 2]
```diff
@@ -113,6 +113,19 @@ def add_training_samples(
             batcher.add_training_samples(training_sample_group=group0)  # -> None
             batcher.add_training_samples(training_sample_group=group1)  # -> TrainingBatch
         """
+        training_batch, _ = self.add_training_samples_with_status(
+            training_sample_group=training_sample_group
+        )
+        return training_batch
+
+    def add_training_samples_with_status(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/components/batcher.py]: I don't think we need to introduce a separate function `add_training_samples_with_status`. Instead we can just modify the original `add_training_samples` to return an extra bool.

Right now it's only used in controller and a unit test.

Comment (taking-lying-flat) [reply torchtitan/experiments/rl/components/batcher.py]: Addressed in 615666e27. I removed add_training_samples_with_status and changed the original add_training_samples method to return (TrainingBatch | None, group_is_trainable). The controller now calls that method directly, and the existing unit-test call sites assert the tuple contract. I also merged the latest main and resolved the fold-batch-dim documentation conflict while preserving the new API. The focused async-controller suite passes: 14 tests.

**Review by tianyu-l — APPROVED** [review]
